### PR TITLE
[Java][RestTemplate][WebClient] Templatized query params for metrics and evicts oom

### LINKED ISSUE #7899: [BUG][Java][spring][webclient] webclient metrics contains actual query parameters, resulting in OOM
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

Generate the spring webclient implementation for an open api endpoint which is using query parameters.
Use the generated ApiClient to make calls to the endpoint with different query parameters in a spring webflux application (with micrometer metrics).

Check the uri tag for the metric http_client_requests.
This uri tag contains the query parameter. This will continue to grow as different values are used.
Eventually, leading to an OutOfMemory error.

##### openapi-generator version

openapi-generator version: 4.3.1

This is not a regression.

##### OpenAPI declaration file content or url

This can be reproduced with the sample petstore fake endpoints yaml file in the project

[/fake/body-with-query-params](https://github.com/OpenAPITools/openapi-generator/blob/9377dbca56afcc32901262e94ba4d7af4379ba94/modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml#L939-L949)

##### Steps to reproduce

1. Generate the spring webclient implementation for the yaml
2. Use it in a Spring boot project with actuator and metrics
3. Call the endpoint with query parameters

Actual

- Query parameter values are present in the http_client_requests metric
Screenshot from 2020-11-08 13-17-35

Expected

- Query parameters values *must* not be present in the metrics.

##### Suggest a fix

Solution for this is available in [spring framework 5.2 webclient](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/reactive/function/client/WebClient.UriSpec.html#uri-java.lang.String-java.util.function.Function-)

I could provide a PR, but that means the minimum required version of spring framework will be 5.2.
Or, we need to make rules in the template to add the following code on spring 5.2+ only, hence leaving it for you to decide.

Here is the fix to the ApiClient.mustache template required to fix this. [webclient-ApiClient.patch.txt](https://github.com/OpenAPITools/openapi-generator/files/5506223/webclient-ApiClient.patch.txt)

```diff
diff --git a/modules/openapi-generator/src/main/resources/Java/libraries/webclient/ApiClient.mustache b/modules/openapi-generator/src/main/resources/Java/libraries/webclient/ApiClient.mustache
index 98f4bfb2..080b6fbd 100644
--- a/modules/openapi-generator/src/main/resources/Java/libraries/webclient/ApiClient.mustache
+++ b/modules/openapi-generator/src/main/resources/Java/libraries/webclient/ApiClient.mustache
@@ -564,11 +564,13 @@ public class ApiClient{{#java8}} extends JavaTimeFormatter{{/java8}} {
         updateParamsForAuth(authNames, queryParams, headerParams, cookieParams);
 
         final UriComponentsBuilder builder = UriComponentsBuilder.fromHttpUrl(basePath).path(path);
-        if (queryParams != null) {
-            builder.queryParams(queryParams);
-        }
 
-        final WebClient.RequestBodySpec requestBuilder = webClient.method(method).uri(builder.build(false).toUriString(), pathParams);
+        final WebClient.RequestBodySpec requestBuilder = webClient.method(method).uri(builder.build(false).toUriString(), uriBuilder -> {
+            if (queryParams == null) {
+                return uriBuilder.build(pathParams);
+            }
+            return uriBuilder.queryParams(queryParams).build(pathParams);
+        });
         if(accept != null) {
             requestBuilder.accept(accept.toArray(new MediaType[accept.size()]));
         }
```

### PR DESCRIPTION
**Pull request by jorgerod**
For both **RestTemplate** and **WebClient** java clients were not being query params

This causes uriTags to be created for each query parameter and could lead to an Out Of Memory.

This PR Resolves #7899 and extends the feature for RestTemplate.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request. CC/ @bbdouglas @sreeshas @jfiala @lukoyanov @cbornet @jeff9finger @karismann @Zomzog @lwlee2608 @nmuesch

### CODE REVIEW

**Comment by jorgerod** [comment]
Hi


I see the next error in _continous-integration/appveyor/pr_ 

```java
  Stack Trace:
     at Org.OpenAPITools.Api.PetApi.UpdatePetWithFormWithHttpInfoAsync(Int64 petId, String name, String status, CancellationToken cancellationToken) in C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient-httpclient\src\Org.OpenAPITools\Api\PetApi.cs:line 1680
   at Org.OpenAPITools.Api.PetApi.UpdatePetWithFormAsync(Int64 petId, String name, String status, CancellationToken cancellationToken) in C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient-httpclient\src\Org.OpenAPITools\Api\PetApi.cs:line 1623
   at Org.OpenAPITools.Test.Api.PetApiTestsV2.UpdatePetWithFormAsync_GivenExistentId_UpdatesTheFields() in C:\projects\openapi-generator\samples\client\petstore\csharp-netcore\OpenAPIClient-httpclient\src\Org.OpenAPITools.Test\Api\PetApiTestsV2.cs:line 442
```

but i have not changed anything related with csharp-netcore.
Can you help me? @bbdouglas @sreeshas @jfiala @lukoyanov @cbornet @jeff9finger @karismann @Zomzog @lwlee2608 @nmuesch @wing328 

thank you

File: modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/ApiClient.mustache [hunk 1]
```diff
@@ -595,13 +596,41 @@ public class ApiClient{{#jsr310}} extends JavaTimeFormatter{{/jsr310}} {
     public String expandPath(String pathTemplate, Map<String, Object> variables) {
         return restTemplate.getUriTemplateHandler().expand(pathTemplate, variables).toString();
     }
+    
+    private String generateQueryUri(MultiValueMap<String, String> queryParams, Map<String, Object> uriParams) {
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/ApiClient.mustache:600]: Suggestion: adding a docstring to explaining what this function does

File: modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/ApiClient.mustache [hunk 2]
```diff
@@ -613,25 +642,22 @@ public class ApiClient{{#jsr310}} extends JavaTimeFormatter{{/jsr310}} {
...
-                        try {
-                            values.set(i, URLEncoder.encode(values.get(i), "utf8"));
-                        } catch (UnsupportedEncodingException e) {
-
-                        }
-                    }
-                }
-            }
-            builder.queryParams(queryParams);
+   	    Map<String,Object> uriParams = new HashMap();
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/ApiClient.mustache:648]: suggestion: remove additional 2-space at the start of the line

Comment (jorgerod) [reply modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/ApiClient.mustache:648]: @wing328 I don't see what space needs to be removed at the beginning of the line. I see it correctly

File: modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/ApiClient.mustache (hunk 1 above)
Comment (jorgerod) [reply modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/ApiClient.mustache:606]: Done

File: modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/ApiClient.mustache (hunk 2 above)
Comment (wing328) [reply modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/ApiClient.mustache:654]: You're right. My mistake.

**Comment by jorgerod** [comment]
Hi @wing328 

I don't understand why failed _continuous-integration/appveyor/pr — AppVeyor build failed_
Can you help me?

**Comment by wing328** [comment]
Appveyor failure not related to this change.

**Comment by admoca60** [comment]
Thank you and congratulations @jorgerod !

File: modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/api.mustache [hunk 3]
```diff
@@ -144,8 +143,8 @@ public class {{classname}} {
         String[] authNames = new String[] { {{#authMethods}}"{{name}}"{{^-last}}, {{/-last}}{{/authMethods}} };
 
         {{#returnType}}ParameterizedTypeReference<{{{returnType}}}> returnType = new ParameterizedTypeReference<{{{returnType}}}>() {};{{/returnType}}{{^returnType}}ParameterizedTypeReference<Void> returnType = new ParameterizedTypeReference<Void>() {};{{/returnType}}
-        return apiClient.invokeAPI(path, HttpMethod.{{httpMethod}}, queryParams, postBody, headerParams, cookieParams, formParams, localVarAccept, localVarContentType, authNames, returnType);
-    }
+        return apiClient.invokeAPI("{{{path}}}", HttpMethod.{{httpMethod}}, {{#hasPathParams}}uriVariables{{/hasPathParams}}{{^hasPathParams}}Collections.<String, Object>emptyMap(){{/hasPathParams}}, queryParams, postBody, headerParams, cookieParams, formParams, localVarAccept, localVarContentType, authNames, returnType);
+      }
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/api.mustache:147]: Looks like 2 additional spaces is added in this line. I'll file a PR to fix it later.

Comment (jorgerod) [reply modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/api.mustache:147]: Ok, thank you

Comment (wing328) [reply modules/openapi-generator/src/main/resources/Java/libraries/resttemplate/api.mustache:147]: Submitted https://github.com/OpenAPITools/openapi-generator/pull/10033

**Pull request merged**